lib/linked_list.rb

```diff
@@ -3,43 +3,68 @@
 class LinkedList
   attr_reader :head
 
-  def initialize
+  def initialize(nextNode = nil)
     @head = nil
+    @next = nextNode
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1), we will always be adding to the beginning of the list
+  # Space complexity - O(1)
   def add_first(data)
-
+    @head = Node.new(data, @head)
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1), only looking at @head
+  # Space complexity - O(1)
   def get_first
-
+    return @head ? @head.data : nil
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n), where n is the number of nodes in the list
+  # Space complexity - O(1)
   def length
-    return 0
+    return 0 if !@head
+    
+    current = @head
+    count = 0 
+    until !current
+      count += 1
+      current = current.next
+    end
+
+    return count
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n), where n is the number of nodes in the list
+  # Space complexity - O(1)
   def add_last(data)
-
+    return add_first(data) if !@head
+    
+    current = @head
+    current = current.next until !current.next
+    current.next = Node.new(data)
   end
 
-  # Time complexity - ?
+  # Time complexity - O(n), where n is the number of nodes in the list
   # Space complexity - ?
   def get_last
```

Comment:
Question, what if `@head` is nil?

Reply:
Oops, good point! I needed to put `return nil if !@head` at the beginning, above line 51. Thanks Chris!